### PR TITLE
feat(experiment): Added Litmus experiment for deploying OpenEBS zfs pv provisioner

### PR DESCRIPTION
**Pull request by w3aman**
Signed-off-by: Aman Gupta <aman.gupta@mayadata.io>

**What this PR does / why we need it**:

- This experiment creates zpool on each node of k8s cluster
- pre-requisite is zfs utils must be pre installed
- it create zpool and storage class for the pool

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Checklist**

* [ ] Does this PR have a corresponding GitHub issue?
* [ ] Have you included relevant README for the chaoslib/experiment with details?
* [ ] Have you added debug messages where necessary? 
* [ ] Have you added task comments where necessary? 
* [ ] Have you tested the changes for possible failure conditions?
* [ ] Have you provided the positive & negative test logs for the litmusbook execution?
* [ ] Does the litmusbook ensure idempotency of cluster state?, i.e., is cluster restored to original state?
* [ ] Have you used non-shell/command modules for Kubernetes tasks?
* [ ] Have you (jinja) templatized custom scripts that is run by the litmusbook, if any? 
* [ ] Have you (jinja) templatized Kubernetes deployment manifests used by the litmusbook, if any?
* [ ] Have you reused/created util functions instead of repeating tasks in the litmusbook?
* [ ] Do the artifacts follow the appropriate directory structure? 
* [ ] Have you isolated storage (eg: OpenEBS) specific implementations, checks? 
* [ ] Have you isolated platform (eg: baremetal kubeadm/openshift/aws/gcloud) specific implementations, checks?  
* [ ] Are the ansible facts well defined? Is the scope explicitly set for playbook & included utils?
* [ ] Have you ensured minimum/careful usage of shell utilities (awk, grep, sed, cut, xargs etc.,)?
* [ ] Can the limtusbook be executed both from within & outside a container (configurable paths, no hardcode)?
* [ ] Can you suggest the minimal resource requirements for the litmusbook execution?
* [ ] Does the litmusbook job artifact carry comments/default options/range for the ENV tunables?
* [ ] Has the litmusbooks been linted? 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by w3aman** [comment]
PLAY [localhost] ***************************************************************
2020-01-30T12:32:57.000762 (delta: 0.116272)         elapsed: 0.116272 ******** 
=============================================================================== 

TASK [Gathering Facts] *********************************************************
2020-01-30T12:32:57.022494 (delta: 0.021672)         elapsed: 0.138004 ******** 
ok: [127.0.0.1]

TASK [include_tasks] ***********************************************************
2020-01-30T12:33:06.536892 (delta: 9.51435)         elapsed: 9.652402 ********* 
included: /utils/fcm/create_testname.yml for 127.0.0.1

TASK [Record test instance/run ID] *********************************************
2020-01-30T12:33:06.691090 (delta: 0.154128)         elapsed: 9.8066 ********** 
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Construct testname appended with runID] **********************************
2020-01-30T12:33:06.777536 (delta: 0.086363)         elapsed: 9.893046 ******** 
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [include_tasks] ***********************************************************
2020-01-30T12:33:06.876876 (delta: 0.099204)         elapsed: 9.992386 ******** 
included: /utils/fcm/update_litmus_result_resource.yml for 127.0.0.1

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
2020-01-30T12:33:07.028110 (delta: 0.151163)         elapsed: 10.14362 ******** 
changed: [127.0.0.1] => {"changed": true, "checksum": "bd09d0129f90d057dbbfd1181c241822e61176ba", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "2f7b795ae84f10e5f9a791b4899b4bc6", "mode": "0644", "owner": "root", "size": 418, "src": "/root/.ansible/tmp/ansible-tmp-1580387587.11-23803040181712/source", "state": "file", "uid": 0}

TASK [Analyze the cr yaml] *****************************************************
2020-01-30T12:33:08.082532 (delta: 1.05426)         elapsed: 11.198042 ******** 
changed: [127.0.0.1] => {"changed": true, "cmd": "cat litmus-result.yaml", "delta": "0:00:00.907506", "end": "2020-01-30 12:33:09.452146", "rc": 0, "start": "2020-01-30 12:33:08.544640", "stderr": "", "stderr_lines": [], "stdout": "---\napiVersion: litmus.io/v1alpha1\nkind: LitmusResult\nmetadata:\n\n  # name of the litmus testcase\n  name: zfspv-provisioner \nspec:\n\n  # holds information on the testcase\n  testMetadata:\n    app:  \n    chaostype:  \n\n  # holds the state of testcase,  manually updated by json merge patch\n  # result is the useful value today, but anticipate phase use in future \n  testStatus: \n    phase: in-progress  \n    result: none ", "stdout_lines": ["---", "apiVersion: litmus.io/v1alpha1", "kind: LitmusResult", "metadata:", "", "  # name of the litmus testcase", "  name: zfspv-provisioner ", "spec:", "", "  # holds information on the testcase", "  testMetadata:", "    app:  ", "    chaostype:  ", "", "  # holds the state of testcase,  manually updated by json merge patch", "  # result is the useful value today, but anticipate phase use in future ", "  testStatus: ", "    phase: in-progress  ", "    result: none "]}

TASK [Apply the litmus result CR] **********************************************
2020-01-30T12:33:09.543696 (delta: 1.461081)         elapsed: 12.659206 ******* 
changed: [127.0.0.1] => {"changed": true, "method": "patch", "result": {"apiVersion": "litmus.io/v1alpha1", "kind": "LitmusResult", "metadata": {"creationTimestamp": "2020-01-22T11:41:14Z", "generation": 3, "name": "zfspv-provisioner", "resourceVersion": "98250", "selfLink": "/apis/litmus.io/v1alpha1/litmusresults/zfspv-provisioner", "uid": "07cea78b-984f-4e85-9b48-91ea3b16902f"}, "spec": {"testMetadata": {}, "testStatus": {"phase": "in-progress", "result": "none"}}}}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
2020-01-30T12:33:11.179152 (delta: 1.635362)         elapsed: 14.294662 ******* 
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Analyze the cr yaml] *****************************************************
2020-01-30T12:33:11.280289 (delta: 0.101071)         elapsed: 14.395799 ******* 
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Apply the litmus result CR] **********************************************
2020-01-30T12:33:11.370357 (delta: 0.090005)         elapsed: 14.485867 ******* 
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [include_tasks] ***********************************************************
2020-01-30T12:33:11.488089 (delta: 0.117655)         elapsed: 14.603599 ******* 
included: /utils/zpool_creation/create_zpool.yml for 127.0.0.1

TASK [Getting the computing node names] ****************************************
2020-01-30T12:33:11.629637 (delta: 0.141044)         elapsed: 14.745147 ******* 
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get nodes --no-headers | grep -v master | awk '{print $1}'", "delta": "0:00:01.378278", "end": "2020-01-30 12:33:13.286921", "rc": 0, "start": "2020-01-30 12:33:11.908643", "stderr": "", "stderr_lines": [], "stdout": "k8s-node1\nk8s-node2\nk8s-node3", "stdout_lines": ["k8s-node1", "k8s-node2", "k8s-node3"]}

TASK [Create a daemonset with privileged access to create zpool on each node] ***
2020-01-30T12:33:13.421155 (delta: 1.791459)         elapsed: 16.536665 ******* 
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl create -f /utils/zpool_creation/privilege_ds.yml", "delta": "0:00:01.622947", "end": "2020-01-30 12:33:15.393117", "rc": 0, "start": "2020-01-30 12:33:13.770170", "stderr": "", "stderr_lines": [], "stdout": "daemonset.apps/privileged-daemonset-bn5qn created", "stdout_lines": ["daemonset.apps/privileged-daemonset-bn5qn created"]}

TASK [Confirm that the ds pods are running on all nodes] ***********************
2020-01-30T12:33:15.496802 (delta: 2.075554)         elapsed: 18.612312 ******* 
FAILED - RETRYING: Confirm that the ds pods are running on all nodes (20 retries left).
changed: [127.0.0.1] => {"attempts": 2, "changed": true, "cmd": "kubectl get pod -l app=zfs-ds --no-headers -o custom-columns=:status.phase | sort | uniq", "delta": "0:00:01.109479", "end": "2020-01-30 12:33:23.393489", "rc": 0, "start": "2020-01-30 12:33:22.284010", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [List the daemonset pods] *************************************************
2020-01-30T12:33:23.512301 (delta: 8.015416)         elapsed: 26.627811 ******* 
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pods -l app=zfs-ds --no-headers -o custom-columns=:.metadata.name", "delta": "0:00:01.106627", "end": "2020-01-30 12:33:24.878840", "rc": 0, "start": "2020-01-30 12:33:23.772213", "stderr": "", "stderr_lines": [], "stdout": "privileged-daemonset-bn5qn-4jbgw\nprivileged-daemonset-bn5qn-5p6n4\nprivileged-daemonset-bn5qn-hlnfn", "stdout_lines": ["privileged-daemonset-bn5qn-4jbgw", "privileged-daemonset-bn5qn-5p6n4", "privileged-daemonset-bn5qn-hlnfn"]}

TASK [Install ZFS utils and Create ZPOOL on each node when pool type is STRIPE] ***
2020-01-30T12:33:24.984977 (delta: 1.472511)         elapsed: 28.100487 ******* 
changed: [127.0.0.1] => (item=privileged-daemonset-bn5qn-4jbgw) => {"changed": true, "cmd": "kubectl exec -ti privileged-daemonset-bn5qn-4jbgw -- bash -c 'zpool create test-pool /dev/sdc'", "delta": "0:00:01.939942", "end": "2020-01-30 12:33:27.232474", "item": "privileged-daemonset-bn5qn-4jbgw", "rc": 0, "start": "2020-01-30 12:33:25.292532", "stderr": "Unable to use a TTY - input is not a terminal or the right kind of file", "stderr_lines": ["Unable to use a TTY - input is not a terminal or the right kind of file"], "stdout": "", "stdout_lines": []}
changed: [127.0.0.1] => (item=privileged-daemonset-bn5qn-5p6n4) => {"changed": true, "cmd": "kubectl exec -ti privileged-daemonset-bn5qn-5p6n4 -- bash -c 'zpool create test-pool /dev/sdc'", "delta": "0:00:01.868076", "end": "2020-01-30 12:33:29.344364", "item": "privileged-daemonset-bn5qn-5p6n4", "rc": 0, "start": "2020-01-30 12:33:27.476288", "stderr": "Unable to use a TTY - input is not a terminal or the right kind of file", "stderr_lines": ["Unable to use a TTY - input is not a terminal or the right kind of file"], "stdout": "", "stdout_lines": []}
changed: [127.0.0.1] => (item=privileged-daemonset-bn5qn-hlnfn) => {"changed": true, "cmd": "kubectl exec -ti privileged-daemonset-bn5qn-hlnfn -- bash -c 'zpool create test-pool /dev/sdc'", "delta": "0:00:01.783053", "end": "2020-01-30 12:33:31.381409", "item": "privileged-daemonset-bn5qn-hlnfn", "rc": 0, "start": "2020-01-30 12:33:29.598356", "stderr": "Unable to use a TTY - input is not a terminal or the right kind of file", "stderr_lines": ["Unable to use a TTY - input is not a terminal or the right kind of file"], "stdout": "", "stdout_lines": []}

TASK [Install ZFS utils and Create ZPOOL on each node when pool type is MIRROR] ***
2020-01-30T12:33:31.500377 (delta: 6.515246)         elapsed: 34.615887 ******* 
skipping: [127.0.0.1] => (item=privileged-daemonset-bn5qn-4jbgw)  => {"changed": false, "item": "privileged-daemonset-bn5qn-4jbgw", "skip_reason": "Conditional result was False"}
skipping: [127.0.0.1] => (item=privileged-daemonset-bn5qn-5p6n4)  => {"changed": false, "item": "privileged-daemonset-bn5qn-5p6n4", "skip_reason": "Conditional result was False"}
skipping: [127.0.0.1] => (item=privileged-daemonset-bn5qn-hlnfn)  => {"changed": false, "item": "privileged-daemonset-bn5qn-hlnfn", "skip_reason": "Conditional result was False"}

TASK [Install OpenEBS ZFS driver components] ***********************************
2020-01-30T12:33:31.661030 (delta: 0.160566)         elapsed: 34.77654 ******** 
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f https://raw.githubusercontent.com/openebs/zfs-localpv/master/deploy/zfs-operator.yaml", "delta": "0:00:02.329712", "end": "2020-01-30 12:33:34.273767", "rc": 0, "start": "2020-01-30 12:33:31.944055", "stderr": "", "stderr_lines": [], "stdout": "namespace/openebs created\ncustomresourcedefinition.apiextensions.k8s.io/zfsvolumes.openebs.io created\nserviceaccount/openebs-zfs-controller-sa created\nclusterrole.rbac.authorization.k8s.io/openebs-zfs-provisioner-role created\nclusterrolebinding.rbac.authorization.k8s.io/openebs-zfs-provisioner-binding created\nstatefulset.apps/openebs-zfs-controller created\nclusterrole.rbac.authorization.k8s.io/openebs-zfs-attacher-role created\nclusterrolebinding.rbac.authorization.k8s.io/openebs-zfs-attacher-binding created\nclusterrole.rbac.authorization.k8s.io/openebs-zfs-snapshotter-role created\nclusterrolebinding.rbac.authorization.k8s.io/openebs-zfs-snapshotter-binding created\nclusterrole.rbac.authorization.k8s.io/openebs-zfs-cluster-driver-registrar-role created\nclusterrolebinding.rbac.authorization.k8s.io/openebs-zfs-cluster-driver-registrar-binding created\nserviceaccount/openebs-zfs-node-sa created\nclusterrole.rbac.authorization.k8s.io/openebs-zfs-driver-registrar-role created\nclusterrolebinding.rbac.authorization.k8s.io/openebs-zfs-driver-registrar-binding created\ndaemonset.apps/openebs-zfs-node created", "stdout_lines": ["namespace/openebs created", "customresourcedefinition.apiextensions.k8s.io/zfsvolumes.openebs.io created", "serviceaccount/openebs-zfs-controller-sa created", "clusterrole.rbac.authorization.k8s.io/openebs-zfs-provisioner-role created", "clusterrolebinding.rbac.authorization.k8s.io/openebs-zfs-provisioner-binding created", "statefulset.apps/openebs-zfs-controller created", "clusterrole.rbac.authorization.k8s.io/openebs-zfs-attacher-role created", "clusterrolebinding.rbac.authorization.k8s.io/openebs-zfs-attacher-binding created", "clusterrole.rbac.authorization.k8s.io/openebs-zfs-snapshotter-role created", "clusterrolebinding.rbac.authorization.k8s.io/openebs-zfs-snapshotter-binding created", "clusterrole.rbac.authorization.k8s.io/openebs-zfs-cluster-driver-registrar-role created", "clusterrolebinding.rbac.authorization.k8s.io/openebs-zfs-cluster-driver-registrar-binding created", "serviceaccount/openebs-zfs-node-sa created", "clusterrole.rbac.authorization.k8s.io/openebs-zfs-driver-registrar-role created", "clusterrolebinding.rbac.authorization.k8s.io/openebs-zfs-driver-registrar-binding created", "daemonset.apps/openebs-zfs-node created"]}

TASK [Verify that the ZFS-controller pod and zfs-node daemonset is running] ****
2020-01-30T12:33:34.380856 (delta: 2.719737)         elapsed: 37.496366 ******* 
FAILED - RETRYING: Verify that the ZFS-controller pod and zfs-node daemonset is running (10 retries left).
FAILED - RETRYING: Verify that the ZFS-controller pod and zfs-node daemonset is running (9 retries left).
FAILED - RETRYING: Verify that the ZFS-controller pod and zfs-node daemonset is running (8 retries left).
FAILED - RETRYING: Verify that the ZFS-controller pod and zfs-node daemonset is running (7 retries left).
FAILED - RETRYING: Verify that the ZFS-controller pod and zfs-node daemonset is running (6 retries left).
FAILED - RETRYING: Verify that the ZFS-controller pod and zfs-node daemonset is running (5 retries left).
changed: [127.0.0.1] => {"attempts": 7, "changed": true, "cmd": "kubectl get pods -n kube-system -l role=openebs-zfs --no-headers -o custom-columns=:status.phase | sort | uniq", "delta": "0:00:01.141710", "end": "2020-01-30 12:34:14.610154", "rc": 0, "start": "2020-01-30 12:34:13.468444", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [Update the storage class template with the variables.] *******************
2020-01-30T12:34:14.705771 (delta: 40.324843)         elapsed: 77.821281 ****** 
changed: [127.0.0.1] => {"changed": true, "checksum": "bccc4006dc5d7f4a504b9cf92bdff4ed236db8f2", "dest": "./openebs-zfspv-sc.yml", "gid": 0, "group": "root", "md5sum": "799eb23a67f4ea084b52fdf4da984f7d", "mode": "0644", "owner": "root", "size": 258, "src": "/root/.ansible/tmp/ansible-tmp-1580387654.77-189955935626110/source", "state": "file", "uid": 0}

TASK [Create Storageclass] *****************************************************
2020-01-30T12:34:15.211627 (delta: 0.505781)         elapsed: 78.327137 ******* 
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f openebs-zfspv-sc.yml", "delta": "0:00:01.424011", "end": "2020-01-30 12:34:16.863270", "failed_when_result": false, "rc": 0, "start": "2020-01-30 12:34:15.439259", "stderr": "", "stderr_lines": [], "stdout": "storageclass.storage.k8s.io/test-sc created", "stdout_lines": ["storageclass.storage.k8s.io/test-sc created"]}

TASK [set_fact] ****************************************************************
2020-01-30T12:34:16.961524 (delta: 1.749811)         elapsed: 80.077034 ******* 
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [include_tasks] ***********************************************************
2020-01-30T12:34:17.063607 (delta: 0.101999)         elapsed: 80.179117 ******* 
included: /utils/fcm/update_litmus_result_resource.yml for 127.0.0.1

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
2020-01-30T12:34:17.190803 (delta: 0.127109)         elapsed: 80.306313 ******* 
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Analyze the cr yaml] *****************************************************
2020-01-30T12:34:17.265562 (delta: 0.074691)         elapsed: 80.381072 ******* 
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Apply the litmus result CR] **********************************************
2020-01-30T12:34:17.342767 (delta: 0.077137)         elapsed: 80.458277 ******* 
skipping: [127.0.0.1] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
2020-01-30T12:34:17.428633 (delta: 0.085773)         elapsed: 80.544143 ******* 
changed: [127.0.0.1] => {"changed": true, "checksum": "84c075e8334d7bd333751b028a5f663797f93e4c", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "d434e4e3eeb0ec3e8c5b3177ff0f0224", "mode": "0644", "owner": "root", "size": 416, "src": "/root/.ansible/tmp/ansible-tmp-1580387657.51-225627615934496/source", "state": "file", "uid": 0}

TASK [Analyze the cr yaml] *****************************************************
2020-01-30T12:34:19.910718 (delta: 2.481987)         elapsed: 83.026228 ******* 
changed: [127.0.0.1] => {"changed": true, "cmd": "cat litmus-result.yaml", "delta": "0:00:01.058608", "end": "2020-01-30 12:34:21.273923", "rc": 0, "start": "2020-01-30 12:34:20.215315", "stderr": "", "stderr_lines": [], "stdout": "---\napiVersion: litmus.io/v1alpha1\nkind: LitmusResult\nmetadata:\n\n  # name of the litmus testcase\n  name: zfspv-provisioner \nspec:\n\n  # holds information on the testcase\n  testMetadata:\n    app:  \n    chaostype:  \n\n  # holds the state of testcase,  manually updated by json merge patch\n  # result is the useful value today, but anticipate phase use in future \n  testStatus: \n    phase: completed  \n    result: Pass ", "stdout_lines": ["---", "apiVersion: litmus.io/v1alpha1", "kind: LitmusResult", "metadata:", "", "  # name of the litmus testcase", "  name: zfspv-provisioner ", "spec:", "", "  # holds information on the testcase", "  testMetadata:", "    app:  ", "    chaostype:  ", "", "  # holds the state of testcase,  manually updated by json merge patch", "  # result is the useful value today, but anticipate phase use in future ", "  testStatus: ", "    phase: completed  ", "    result: Pass "]}

TASK [Apply the litmus result CR] **********************************************
2020-01-30T12:34:21.368038 (delta: 1.45684)         elapsed: 84.483548 ******** 
changed: [127.0.0.1] => {"changed": true, "failed_when_result": false, "method": "patch", "result": {"apiVersion": "litmus.io/v1alpha1", "kind": "LitmusResult", "metadata": {"creationTimestamp": "2020-01-22T11:41:14Z", "generation": 4, "name": "zfspv-provisioner", "resourceVersion": "98602", "selfLink": "/apis/litmus.io/v1alpha1/litmusresults/zfspv-provisioner", "uid": "07cea78b-984f-4e85-9b48-91ea3b16902f"}, "spec": {"testMetadata": {}, "testStatus": {"phase": "completed", "result": "Pass"}}}}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=21   changed=15   unreachable=0    failed=0   

2020-01-30T12:34:22.776432 (delta: 1.40831)         elapsed: 85.891942 ******** 
===============================================================================

File: providers/zfs-localpv-provisioner/openebs-zfspv-sc.j2 [hunk 1]
```diff
@@ -0,0 +1,13 @@
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: "{{ storage_class }}"
+parameters:
+  recordsize: "4k"
+  compression: "off"
```
Comment (gprasath) [inline providers/zfs-localpv-provisioner/openebs-zfspv-sc.j2]: get all the parameters as env

File: providers/zfs-localpv-provisioner/run_litmus_test.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        env:
+          - name: ANSIBLE_STDOUT_CALLBACK
+            #value: log_plays
+            value: default
+
+            ## Storage class will be created with the name provided here
+          - name: STORAGE_CLASS  
+            value: ''
+
+          - name: POOL_NAME
```
Comment (gprasath) [inline providers/zfs-localpv-provisioner/run_litmus_test.yml:30]: Comments please

File: utils/zpool_creation/privilege_ds.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+---
+  apiVersion: apps/v1
+  kind: DaemonSet
+  metadata:
+    generateName: privileged-daemonset-
```
Comment (gprasath) [inline utils/zpool_creation/privilege_ds.yml]: Change the name - > zpool-creator

Comment (gprasath) [reply utils/zpool_creation/privilege_ds.yml]: Change the file name as well -> create_zpool_ds.yml

File: utils/zpool_creation/privilege_ds.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        app: zfs-ds
+    template:
+      metadata:
+        labels:
+          app: zfs-ds
+      spec:
+        containers:
+        - name: zfsutils
+          image: w3aman/zfsutils:ci
+          imagePullPolicy: Always
```
Comment (gprasath) [inline utils/zpool_creation/privilege_ds.yml]: make it ifnotpresent

File: utils/zpool_creation/create_zpool.yml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+   - name: List the daemonset pods 
+     shell: >
+       kubectl get pods -l app=zfs-ds --no-headers -o custom-columns=:.metadata.name
+     args:
+       executable: /bin/bash
+     register: zfs_ds_pods
+
+   - block:
+        
+      - name: Install ZFS utils and Create ZPOOL on each node when pool type is STRIPE
```
Comment (gprasath) [inline utils/zpool_creation/create_zpool.yml]: Not installing zfs utils here. :)

File: utils/zpool_creation/create_zpool.yml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+          kubectl exec -ti {{ item }} -- bash -c 'zpool create {{ pool_name }} /dev/sdb'
+        args: 
+          executable: /bin/bash
+        with_items: "{{ zfs_ds_pods.stdout_lines }}"
+
+     when: "pool_type == 'striped'"
+
+   - block: 
+     
+      - name: Install ZFS utils and Create ZPOOL on each node when pool type is MIRROR
```
Comment (gprasath) [inline utils/zpool_creation/create_zpool.yml]: check task name

File: utils/zpool_creation/create_zpool.yml [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+   - block:
+        
+      - name: Install ZFS utils and Create ZPOOL on each node when pool type is STRIPE
+        shell: >
+          kubectl exec -ti {{ item }} -- bash -c 'zpool create {{ pool_name }} /dev/sdb'
+        args: 
+          executable: /bin/bash
+        with_items: "{{ zfs_ds_pods.stdout_lines }}"
+
+     when: "pool_type == 'striped'"
```
Comment (gprasath) [inline utils/zpool_creation/create_zpool.yml]: add a check to ensure that the pool is created successfully

File: utils/zpool_creation/create_zpool.yml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+   - block: 
+     
+      - name: Install ZFS utils and Create ZPOOL on each node when pool type is MIRROR
+        shell: >
+          kubectl exec -ti {{ item }} -- bash -c 'zpool create {{ pool_name }} mirror /dev/sdb /dev/sdc'
+        args: 
+          executable: /bin/bash
+        with_items: "{{ zfs_ds_pods.stdout_lines }}"
```
Comment (gprasath) [inline utils/zpool_creation/create_zpool.yml:55]: add a check to ensure that the pool is created successfully

File: utils/zpool_creation/create_zpool.yml [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+     until: "result.stdout == 'Running'"
+     delay: 5
+     retries: 20
+
+   - name: List the daemonset pods 
+     shell: >
+       kubectl get pods -l app=zfs-ds --no-headers -o custom-columns=:.metadata.name
+     args:
+       executable: /bin/bash
+     register: zfs_ds_pods
```
Comment (gprasath) [inline utils/zpool_creation/create_zpool.yml]: So, pools will be created on all the nodes?

**Review by gprasath — CHANGES_REQUESTED** [review]

File: utils/zpool_creation/create_zpool.yml (hunk 8 above)
Comment (w3aman) [reply utils/zpool_creation/create_zpool.yml:55]: Added

Comment (w3aman) [reply utils/zpool_creation/create_zpool.yml:55]: added

File: utils/zpool_creation/create_zpool.yml (hunk 9 above)
Comment (w3aman) [reply utils/zpool_creation/create_zpool.yml]: yes

**Review by gprasath — APPROVED** [review]

**Pull request merged**